smart-contracts/contracts/cl-vault/src/test_tube/asserts.rs

```diff
@@ -0,0 +1,52 @@
+use std::str::FromStr;
+
+use cosmwasm_std::{Decimal, Uint128};
+
+#[macro_export]
+macro_rules! assert_eq_with_diff {
+    ($left:expr, $left_tag:expr, $right:expr, $right_tag:expr, $max_rel_diff:expr, $max_absolute_diff:expr $(,)?) => {{
+        $crate::test_tube::asserts::assert_eq_with_diff_impl($left, $left_tag, $right, $right_tag, $max_rel_diff, $max_absolute_diff, None);
+    }};
+    ($left:expr, $left_tag:expr, $right:expr, $right_tag:expr, $max_rel_diff:expr, $max_absolute_diff:expr, $($args:tt)+) => {{
+        $crate::test_tube::asserts::assert_eq_with_diff_impl($left, $left_tag, $right, $right_tag, $max_rel_diff, $max_absolute_diff, Some(format!($($args)*)));
+    }};
+}
+
+/// assert that some two values are at most both the relative difference and absolute difference away from each other
```

Comment:
Combining checks for relative and absolute difference is not a good idea. What is the underlying issue that is addressed here?

Reply:
This was previously used to ensure certain asserts at low token amounts where the absolute diff was bigger than the allowed relative diff still passed. For example if we assert with 100uosmo and 100uatom and we round down twice, we have a diff of 2%. So with this we can still test for approx 0.001% diffs on certain functions like swap functionality and also test low values

Reply:
https://github.com/quasar-finance/quasar/pull/639#discussion_r1662596165 At this comment I do now think I should have used an comparison with only an allowed absolute diff, will change that